packages/sdk-react/src/core/hooks/useAutosizeGridColumns.ts

```diff
@@ -0,0 +1,137 @@
+import { useLayoutEffect, useRef } from 'react';
+import { flushSync } from 'react-dom';
+
+import { useMoniteContext } from '@/core/context/MoniteContext';
+import { GridColDef, useGridApiRef } from '@mui/x-data-grid';
+
+export const defaultCounterpartColumnWidth = 250;
+const maximumDocumentIdColumnWidth = 200;
+
+export function useAreCounterpartsLoading(
+  rows?: { counterpart_id?: string }[]
+) {
+  const { api } = useMoniteContext();
+  const result = !!api.counterparts.getCounterpartsId.useIsFetching({
+    infinite: false,
+    exact: false,
+    predicate: (query) =>
+      !!rows?.some(
+        ({ counterpart_id }) =>
+          !query.state.isInvalidated &&
+          counterpart_id === query.queryKey[1].path.counterpart_id
+      ),
+  });
+  return !rows ? true : result;
+}
+
+// Adapted from https://mui.com/x/react-data-grid/column-dimensions/#autosizing-asynchronously
+// setTimeout and flushSync are necessary for call order control
+// Docs say:
+// The Data Grid can only autosize based on the currently rendered cells.
+// DOM access is required to accurately calculate dimensions
+export function useAutosizeGridColumns(
+  rows: any,
+  columns: GridColDef[],
+  areCounterpartsLoading: boolean,
+  columnSerializationKey: string
+) {
+  const gridApiRef = useGridApiRef();
+
+  const autoSizePerformed = useRef(false); // useRef instead of useState to avoid re-renders on value changes
+  const serializationKey = 'Monite-DataGridColumns-' + columnSerializationKey;
+
+  // Autosize columns
+  useLayoutEffect(() => {
+    const grid = gridApiRef.current;
+    const timeouts: number[] = [];
+    if (
+      rows && // Do not autosize until rows are loaded
+      rows.length &&
+      grid &&
+      grid.autosizeColumns && // Will be undefined in test environment
+      grid.getRowsCount() && // Make sure the grid was already rendered so that autosizeColumns function will work
+      grid.getCellElement(rows[0].id, grid.getAllColumns()[0].field) && // Make sure the grid was already rendered so that autosizeColumns function will work
+      !areCounterpartsLoading && // Skip autosize until counterparts are loaded
+      !autoSizePerformed.current && // Do not perform autosize twice
```

Comment:
Consider setting `autoSizePerformed.current = true` immediately 

```ts

// Before:
if (!autoSizePerformed.current) {
    // Autosize logic
}

// After:
if (!autoSizePerformed.current && !areCounterpartsLoading) {
    autoSizePerformed.current = true;  // Prevent future executions
    // Autosize logic
}
```

 In that case we avoid the potential for multiple executions of autosizing logic if the effect reruns before the current autosize operation completes

Reply:
autoSizePerformed.current is used for determining if current columns state should be saved. Therefore, this value should be set only after actually performing the autosizing operation